_source/_posts/2022-08-31-state-of-java-project-loom.md

```diff
@@ -0,0 +1,261 @@
+
+So in a thread-per-request model, the throughput will be limited by the number of OS threads available, which depends on the number of physical cores/threads available on the hardware. To work around this, you have to use shared thread pools or asynchronous concurrency, both of which have their drawbacks. Thread pools have many limitations, like thread leaking, deadlocks, resource thrashing, etc. Asynchronous concurrency means you must adapt to a more complex programming style and handle data races carefully. There are also chances of memory leaks, thread locking, etc.
+
+Another common use case is parallel processing or multi-threading. You must ensure thread synchronization when executing a parallel task distributed over multiple threads to avoid data races. The implementation becomes even more fragile and puts a lot more responsibility on the developer to ensure there are no issues like thread leaks and cancellation delays.
+
+Project Loom aims to fix these issues in the current concurrency model by introducing two new features: virtual threads and structured concurrency.
+
+### Virtual Threads
+
+> Java 19 is scheduled to be released in September 2022, and Virtual threads will be a preview feature. Yayyy!
+
+[Virtual threads](https://openjdk.org/jeps/425) are lightweight threads that are not tied to OS threads but are managed by the JVM. They are suitable for thread-per-request programming styles without having the limitations of OS threads. You can create millions of virtual threads without affecting throughput. This is quite similar to coroutines, like [Goroutines](https://go.dev/tour/concurrency/1), made famous by the Go programming language.
+
+The new virtual threads in Java 19 will be pretty easy to use. Compare the below with Golang's Goroutines or Kotlin's coroutines.
+
+**Virtual thread**
+
+```java
+Thread.startVirtualThread(() -> {
+    System.out.println("Hello, Project Loom!");
+});
+```
+
+**Goroutine**
+
+```go
+go func() {
+    println("Hello, Goroutines!")
+}()
+```
+
+**Kotlin coroutine**
+
+```kotlin
+runBlocking {
+    launch {
+        println("Hello, Kotlin coroutines!")
+    }
+}
+```
+
+A fun fact is that before JDK 1.1, Java had support for green threads (a.k.a virtual threads), but the feature was removed in JDK 1.1 as that implementation was not any better than platform threads. The new implementation of virtual threads is done in the JVM, where it maps multiple virtual threads to one or more OS threads, and the developer can use virtual threads or platform threads as per their needs. A few other important aspects of virtual threads are:
```

Comment:
```suggestion
Fun fact:  before JDK 1.1, Java had support for green threads (aka virtual threads), but the feature was removed in JDK 1.1 as that implementation was not any better than platform threads. The new implementation of virtual threads is done in the JVM, where it maps multiple virtual threads to one or more OS threads, and the developer can use virtual threads or platform threads as per their needs. A few other important aspects of virtual threads are:
```
i did a little online research, ;-)  - seems like aka without periods is the more popular/recommended usage)